README.md
# NAME

InfluxDB::Writer - Various tools to send lines to InfluxDB

# VERSION

version 1.003

# DESCRIPTION

`InfluxDB::Writer` is a collection of tools & modules to send [lines](https://influxdb.com/docs/v0.9/write_protocols/line.html) to [InfluxDB](https://influxdb.com)

Currently, the suggested setup is to have your apps write stats into a file containing the PID of the process. This file (or files) can then be processed by various tools provided by `InfluxDB::Writer` and their contents sent to InfluxDB.

- [InfluxDB::Writer::FileTailer](https://metacpan.org/pod/InfluxDB%3A%3AWriter%3A%3AFileTailer)

    A daemon using `IO::Async` to watch a directory. All files in this directory that include the PID of a currently running process are tailed and any new lines added sent to InfluxDB in batches. New files added to the directory are also automatically tailed.

- [InfluxDB::Writer::SendLines](https://metacpan.org/pod/InfluxDB%3A%3AWriter%3A%3ASendLines)

    A script that takes one file and send the lines to InfluxDB in batches.

- [InfluxDB::Writer::CompactFiles](https://metacpan.org/pod/InfluxDB%3A%3AWriter%3A%3ACompactFiles)

    A script that checks all files in a directory. All files that contain PIDs of not longer running processes are compacted into one file which is then `gzip`ed.

Please refer to the documentation of the respective modules for detailed information.

# EXAMPLE USAGE

We use [Measure::Everything::InfluxDB::](https://metacpan.org/pod/Measure%3A%3AEverything%3A%3AInfluxDB%3A%3A)

# AUTHOR

Thomas Klausner <domm@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2015 - 2021 by Thomas Klausner.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

# POD ERRORS

Hey! **The above document had some coding errors, which are explained below:**

- Around line 39:

    Unterminated L<...> sequence


Changes
Changes for InfluxDB::Writer

{{$NEXT}}

1.003     2021-01-08 21:58:11+01:00
    - dist housekeeping, Dist::Zilla::PluginBundle::Author::DOMM


1.002   2016-02-01T12:43:58
     - Check /proc/$pid/fd for open file (Klaus Ita)
     - Separate Buffer and to_Send (Klaus Ita)


1.001   2015-12-04T10:59:06
    - File Tailer that archives (and uploads) expired files (koki)
    - Extract Auth to Role (koki)


1.000   2015-09-24T00:15:06
    - init dzil


lib/InfluxDB/Writer.pm
package InfluxDB::Writer;

# ABSTRACT: Various tools to send lines to InfluxDB
# VERSION

use strict;
use warnings;

__END__

=head1 DESCRIPTION

C<InfluxDB::Writer> is a collection of tools & modules to send L<lines|https://influxdb.com/docs/v0.9/write_protocols/line.html> to L<InfluxDB|https://influxdb.com>

Currently, the suggested setup is to have your apps write stats into a file containing the PID of the process. This file (or files) can then be processed by various tools provided by C<InfluxDB::Writer> and their contents sent to InfluxDB.

=over

=item L<InfluxDB::Writer::FileTailer>

A daemon using C<IO::Async> to watch a directory. All files in this directory that include the PID of a currently running process are tailed and any new lines added sent to InfluxDB in batches. New files added to the directory are also automatically tailed.

=item L<InfluxDB::Writer::SendLines>

A script that takes one file and send the lines to InfluxDB in batches.

=item L<InfluxDB::Writer::CompactFiles>

A script that checks all files in a directory. All files that contain PIDs of not longer running processes are compacted into one file which is then C<gzip>ed.

=back

Please refer to the documentation of the respective modules for detailed information.

=head1 EXAMPLE USAGE

We use L<Measure::Everything::InfluxDB::





lib/InfluxDB/Writer/SendLines.pm
package InfluxDB::Writer::SendLines;

# ABSTRACT: Send lines from a file to InfluxDB
# VERSION

use strict;
use warnings;
use feature 'say';


use Moose;
use Carp qw(croak);
use Log::Any qw($log);
use File::Spec::Functions;
use Hijk ();

has 'file'        => ( is => 'ro', isa => 'Str', required => 1 );
has 'influx_host' => ( is => 'ro', isa => 'Str', required => 1 );
has 'influx_port' =>
    ( is => 'ro', isa => 'Int', default => 8086, required => 1 );
has 'influx_db'       => ( is => 'ro', isa => 'Str', required => 1 );
has 'buffer_size'     => ( is => 'ro', isa => 'Int', default  => 1000 );

with qw(InfluxDB::Writer::AuthHeaderRole);

$| = 1;

my @buffer;

sub run {
    my $self = shift;

    $log->infof( "Starting %s with file %s", __PACKAGE__, $self->file );

    my $f     = $self->file;
    my $lines = `wc -l $f`;
    chomp($lines);
    $lines =~ s/ .*//;
    my $total = $lines;
    my $start = scalar time;

    open( my $in, "<", $self->file ) || die $!;

    my $cnt       = 0;
    my $print_cnt = $self->buffer_size * 50;
    while ( my $line = <$in> ) {
        my $elements = push( @buffer, $line );
        if ( $elements >= $self->buffer_size ) {
            $self->send;
        }
        $cnt++;
        if ( $cnt % $print_cnt == 0 ) {
            my $now   = scalar time;
            my $diff  = $now - $start || 1;
            my $speed = $cnt / $diff;
            my $estimate =
                $speed > 0 ? ( $total - $cnt ) / $speed : 'Infinity';
            printf( "  % 6i/%i (%.2f/s) time left: %i sec\n",
                $cnt, $total, $speed, $estimate );
        }
    }
    $self->send;
}

sub send {
    my $self       = shift;
    my $second_try = shift;
    my $new_buffer = shift;

    my $to_send = $second_try ? $new_buffer : \@buffer;

    my %args;
    if ( $self->_with_auth ) {
        $args{head} = [ "Authorization" => $self->_auth_header ];
    }

    $log->debugf( "Sending %i lines to influx", scalar @$to_send );
    my $res = Hijk::request(
        {   method       => "POST",
            host         => $self->influx_host,
            port         => $self->influx_port,
            path         => "/write",
            query_string => "db=" . $self->influx_db,
            body         => join( '', @$to_send ),
            %args,
        }
    );

    if ( $res->{status} != 204 ) {
        if (!$second_try
            && (
                (   exists $res->{error}
                    && $res->{error} & Hijk::Error::TIMEOUT
                )
                || ( $res->{status} == 500 && $res->{body} =~ /timeout/ )
            )
            ) {
            # wait a bit and try again with smaller packages
            my @half = splice( @$to_send, 0, int( scalar @$to_send / 2 ) );
            print ':';
            $self->send( 1, \@half );
            $self->send( 1, $to_send );
        }
        else {
            $log->errorf(
                "Could not send %i lines to influx: %s",
                scalar @$to_send,
                $res->{body}
            );
            open( my $fh, ">>", $self->file . '.err' ) || die $!;
            print $fh join( '', @$to_send );
            close $fh;
            print 'X';
        }
    }
    else { # success
        print $second_try ? ',' : '.';
    }

    @$to_send = () unless $second_try;
}

__PACKAGE__->meta->make_immutable;
1;


lib/InfluxDB/Writer/CompactFiles.pm
package InfluxDB::Writer::CompactFiles;

# ABSTRACT: Collect and compact files containing InfluxDB lines
# VERSION

use strict;
use warnings;
use feature 'say';

use Moose;
use Carp qw(croak);
use Log::Any qw($log);
use File::Spec::Functions;
use Sys::Hostname qw(hostname);
use InfluxDB::LineProtocol qw(line2data data2line);
use Time::Moment;

has 'dir'    => ( is => 'ro', isa => 'Str',     required  => 1 );
has 'tags'   => ( is => 'ro', isa => 'HashRef', predicate => 'has_tags' );
has 'delete' => ( is => 'ro', isa => 'Bool',    default   => 1 );

sub run {
    my $self = shift;

    unless ( -d $self->dir ) {
        croak "Not a directory: " . $self->dir;
    }

    my $now = Time::Moment->now->to_string;
    my $outfile = join('_',hostname(),'stats',$now) . '.compacted';

    my $target = catfile( $self->dir, $outfile );
    open( my $out, ">>", $target ) || die $!;
    $log->infof( "Starting %s of directory %s into %s", __PACKAGE__,
        $self->dir, $target );

    opendir( my $dh, $self->dir );
    while ( my $file = readdir($dh) ) {
        next unless $file =~ /\.stats$/;
        $file =~ /(\d+)\.stats/;
        my $pid = $1;
        my $is_running = kill 0, $pid;
        if ($is_running) {
            $log->debugf( "Skip file %s because pid %i is still running",
                $file, $pid );
            next;
        }
        else {
            $log->infof( "Append file %s to %s", $file, $outfile );
            my $source = catfile( $self->dir, $file );
            open( my $fh, "<", $source );
            while ( my $line = <$fh> ) {
                if ( $self->has_tags ) {
                    $line = $self->add_tags_to_line($line);
                }
                say $out $line;
            }
            if ( $self->delete ) {
                unlink($source) || die "$!";
            }
        }
    }
    system('gzip', $target);
}

sub add_tags_to_line {
    my ( $self, $line ) = @_;

    my ( $measurement, $values, $tags, $timestamp ) = line2data($line);
    my $combined_tags;
    if ($tags) {
        $combined_tags = { %$tags, %{ $self->tags } };
    }
    else {
        $combined_tags = $tags;
    }
    return data2line( $measurement, $values, $combined_tags, $timestamp );
}

__PACKAGE__->meta->make_immutable;
1;


lib/InfluxDB/Writer/AuthHeaderRole.pm
package InfluxDB::Writer::AuthHeaderRole;

# ABSTRACT: Helper role
# VERSION

use Moose::Role;
use MIME::Base64 qw/encode_base64/;
use Log::Any qw($log);

has 'influx_username' => ( is => 'ro', isa => 'Str', required => 0 );
has 'influx_password' => ( is => 'ro', isa => 'Str', required => 0 );

has '_with_auth' => ( is => 'rw', isa => 'Bool', lazy => 1, default => sub { return shift->_auth_header ? 1 : 0 } );
has '_auth_header' => ( is => 'ro', isa => 'Str', lazy_build => 1, builder => '_build__auth_header', required => 1 );

sub _build__auth_header {
    my $self = shift;

    if ( $self->influx_username && $self->influx_password ) {
        my $base64 = encode_base64(
            join( ":", $self->influx_username, $self->influx_password ) );
        chomp($base64);

        return "Basic $base64";
    } else {
        return "";
    }
}

1;


lib/InfluxDB/Writer/RememberingFileTailer.pm
package InfluxDB::Writer::RememberingFileTailer;

# ABSTRACT: RememberingFileTailer
# VERSION

use Moose;
use feature 'say';

use IO::Async::File;
use IO::Async::FileStream;
use IO::Async::Loop;
use Hijk ();
use Carp qw(croak);
use InfluxDB::LineProtocol qw(line2data data2line);
use Log::Any qw($log);
use File::Spec::Functions;
use File::Spec qw(splitpath);

extends 'InfluxDB::Writer::FileTailer';

has 'done_dir' => ( is => 'rw', isa => "Str" );

before 'run' => sub {
    my $self = shift;

    my $done_dir = catdir( $self->dir, 'done' );
    if ( -d $done_dir ) {
        $self->done_dir($done_dir);
    }
    else {
        croak "Missing 'done' directory, please create: " . $done_dir;
    }

};

sub archive_file {
    my $self = shift;
    my $file = shift;

    my $done_dir = $self->done_dir;
    my ( $vol, $dirs, $basename ) = File::Spec->splitpath($file);

    my $destination = catfile( $done_dir, $basename );
    my $count = 0;
    while ( -f $destination ) {
        $count++;
        my $new_destination = catfile( $done_dir, $basename ) . '_' . $count;
        $log->infof( "Renaming destination %s to %s",
            $destination, $new_destination );
        $destination = $new_destination;
    }

    if ( -s $file == 0 ) {

        if ( unlink($file) ) {
            $log->infof( "Unlinked zero size stats file %s", $file );
            return 1;
        }
        else {
            $log->warnf( "Could not unlink zero size stats file %s (%s)",
                $file, $! );
            return 0;
        }

        return 0;
    }

    if ( rename( $file, $destination ) ) {
        $log->infof( "Archived file %s to %s", $basename, $done_dir );
        return 1;
    }
    else {
        $log->errorf( "Failed to archive %s to %s", $basename, $done_dir );
        croak "Failed to archive $file";
        return 0;
    }

}

=head2 slurp_and_send

IO::Async read operations are blocking so it does not make sense to wrice
complicated async code here. We want to read the files that are left over and
send them to influx asap, then move them to the archive folder (aka. out of the
way)

=cut

sub slurp_and_send {
    my $self = shift;
    my $file = shift;

    if ( open( my $fh, "<", $file ) ) {
        $log->infof( "Slurping %s", $file );

        while ( my $line = <$fh> ) {
            chomp($line);
            if ( $self->has_tags ) {
                $line = $self->add_tags_to_line($line);
            }
            if ($line !~ /\s\d+$/) {
                # Line does not end in timestamp
                $log->warnf("Skipping probably broken line %s", $line);
                next;
            
            }
            $self->buffer_push($line);

            if ( $self->buffer_size > $self->flush_size ) {
                if ( !$self->send ) {
                    $log->warnf( "Unable to send buffer (%i lines)",
                        $self->buffer_size );
                    return;
                }
            }
        }

        if ( $self->buffer_size ) {
            $log->infof( "Clear buffer (size %i) for file %s",
                $self->buffer_size, $file );
            if ( !$self->send ) {
                $log->warnf( "Unable to send clear buffer (%i lines)",
                    $self->buffer_size );
                return;
            }
        }

        $log->infof( "Finished slurping %s", $file );

        return 1;
    }

    return;

}

override 'archive_hook' => sub {
    shift->archive_file(@_);
    super();
};

override 'cleanup_hook' => sub {
    my ( $self, $file ) = @_;

    if ( $self->slurp_and_send($file) ) {
        $self->archive_file($file);
    }
    return super();

};

1;


lib/InfluxDB/Writer/FileTailer.pm
package InfluxDB::Writer::FileTailer;

# ABSTRACT: Tail files and send lines to InfluxDB
# VERSION

use strict;
use warnings;
use feature 'say';

use Moose;
use IO::Async::File;
use IO::Async::FileStream;
use IO::Async::Loop;
use Hijk ();
use Carp qw(croak);
use InfluxDB::LineProtocol qw(line2data data2line);
use Log::Any qw($log);
use File::Spec::Functions;
use Cwd 'abs_path';

with qw(InfluxDB::Writer::AuthHeaderRole);

has 'dir'         => ( is => 'ro', isa => 'Str', required => 1 );
has 'influx_host' => ( is => 'ro', isa => 'Str', required => 1 );
has 'influx_port' =>
    ( is => 'ro', isa => 'Int', default => 8086, required => 1 );
has 'influx_db' => ( is => 'ro', isa => 'Str', required => 1 );

has 'flush_size' =>
    ( is => 'ro', isa => 'Int', required => 1, default => 1000 );
has 'flush_interval' =>
    ( is => 'ro', isa => 'Int', required => 1, default => 30 );
has 'tags' => ( is => 'ro', isa => 'HashRef', predicate => 'has_tags' );
has '_files' => ( is => 'ro', isa => 'HashRef', default => sub { {} } );
has '_loop' => ( is => 'ro', isa => 'IO::Async::Loop', lazy_build => 1 );
has 'buffer' => ( is => 'ro', isa => 'ArrayRef[Str]', default => sub { [] }, traits => ['Array'],
    handles => {
        buffer_push => 'push',
        buffer_all => 'elements',
        buffer_size => 'count',
        buffer_splice => 'splice',
        buffer_is_empty => 'is_empty',
    },

);

sub _build__loop {
    return IO::Async::Loop->new;
}

sub run {
    my $self = shift;

    unless ( -d $self->dir ) {
        croak "Not a directory: " . $self->dir;
    }

    $log->infof( "Starting %s in directory %s", __PACKAGE__, $self->dir );

    $self->watch_dir;

    my $dir = IO::Async::File->new(
        filename         => $self->dir,
        on_mtime_changed => sub {
            $self->watch_dir;
        },
    );

    $self->_loop->add($dir);

    my $timer = IO::Async::Timer::Periodic->new(    # could be Countdown
        interval => $self->flush_interval,
        on_tick  => sub {
            $self->send;
        },
    );
    $timer->start;
    $self->_loop->add($timer);

    $self->_loop->run;
}


sub cleanup_hook {}
sub archive_hook {}

sub watch_dir {
    my ($self) = @_;

    $log->infof( "Checking for new files to watch in %s", $self->dir );
    opendir( my $dh, $self->dir );
    while ( my $f = readdir($dh) ) {
        next unless $f =~ /\.stats$/;
        if ( my $watcher =
            $self->setup_file_watcher( catfile( $self->dir, $f ) ) ) {
            $self->_loop->add($watcher);
        }
    }
    closedir($dh);
}

sub is_running {
    my ($self, $pid, $file) = @_;

    my $running = kill('ZERO', $pid);
    return unless $running;

    # This might only work on linux with $proc access
    my $fd_dir = catdir("/", "proc", $pid, "fd");
    if (-d $fd_dir ) {
        my $abs_file = File::Spec->rel2abs( $file );

        my $found = 0;
        opendir(my $dh, $fd_dir);
        while ( my $f = readdir($dh) ) {
            $f = catfile($fd_dir, $f);

            if ( -f $f && -l $f ) {

                my $filename = Cwd::abs_path( $f );
                if ($abs_file eq $filename) { #need to close dir
                    $found = 1;
                    last;
                }
            }
        }
        closedir($dh);
    
        return $found;
    }

    return $running;

}

sub setup_file_watcher {
    my ( $self, $file ) = @_;

    $file =~ /(\d+)\.stats$/;
    my $pid = int($1);

    my $is_running = $self->is_running( $pid, $file );

    if (!$is_running) {

        if ( my $w = $self->_files->{$file} ) {
            $self->_loop->remove($w);
            undef $w;
            delete $self->_files->{$file};
            $log->infof( "Removed watcher for %s because pid %i is not more",
                $file, $pid );

            $self->archive_hook($file);

        }
        else {
            $log->debugf(
                "Skipping file %s because pid %i seems to be not running.",
                $file, $pid );

            $self->cleanup_hook($file);
        }
        return;
    }

    if ( $self->_files->{$file} ) {
        $log->debugf( "Already watching file %s", $file );
        return;
    }

    if ( open( my $fh, "<", $file ) ) {
        my $filestream = IO::Async::FileStream->new(
            read_handle => $fh,
            on_initial  => sub {
                my ($stream) = @_;
                $stream->seek_to_last("\n");    # TODO remember last position?
            },

            on_read => sub {
                my ( $stream, $buffref ) = @_;

                while ( $$buffref =~ s/^(.*\n)// ) {
                    my $line = $1;
                    if ( $self->has_tags ) {
                        $line = $self->add_tags_to_line($line);
                    }
                    $self->buffer_push($line);
                }

                if ( $self->buffer_size > $self->flush_size ) {
                    $self->send;
                }

                return 0;
            },
        );
        $log->infof( "Tailing file %s", $file );
        $self->_files->{$file} = $filestream;
        return $filestream;
    }
    else {
        $log->errorf( "Could not open file %s: %s", $file, $! );
        return;
    }
}

sub send {
    my ($self) = @_;

    my $current_size = $self->buffer_size;
    return unless $current_size;

    my @to_send = $self->buffer_splice(0, $current_size);

    my %args;
    if ( $self->_with_auth ) {
        $args{head} = [ "Authorization" => $self->_auth_header ];
    }

    $log->debugf( "Sending %i lines to influx", scalar @to_send);

    ( my $body = join("\n", @to_send) ) =~ s/\n{2,}/\n/gs;

    my $request_data = { 
        method       => "POST",
        host         => $self->influx_host,
        port         => $self->influx_port,
        path         => "/write",
        query_string => "db=" . $self->influx_db,
        body         => $body,
        %args,
    };
    $log->debugf("The Hijk::Request: %s", $request_data);
    my $res = Hijk::request($request_data);

    if (my $current_error = $res->{error}) {

        my @errs = (qw/
            CONNECT_TIMEOUT
            READ_TIMEOUT
            TIMEOUT
            CANNOT_RESOLVE
            REQUEST_SELECT_ERROR
            REQUEST_WRITE_ERROR
            REQUEST_ERROR
            RESPONSE_READ_ERROR
            RESPONSE_BAD_READ_VALUE
            RESPONSE_ERROR
            /);

        my @matches;
        foreach my $err (@errs) {
            my $const = eval "Hijk::Error::" . $err;
            if ( $current_error & $const ) {
                push(@matches, $err);
            }
        }

        $log->errorf("Hijk Request Error(s) cannot send %s", join(", ", @matches));

        return;
    
    }
    if ( $res->{status} != 204 ) {
        $log->errorf(
            "Could not send %i lines to influx: %s, req: %s",
            scalar @to_send,
            $res->{body},
            $request_data
        );

        return;
    }

    return scalar(@to_send);
}

sub add_tags_to_line {
    my ( $self, $line ) = @_;

    my ( $measurement, $values, $tags, $timestamp ) = line2data($line);
    my $combined_tags;
    if ($tags) {
        $combined_tags = { %$tags, %{ $self->tags } };
    }
    else {
        $combined_tags = $tags;
    }
    return data2line( $measurement, $values, $combined_tags, $timestamp );
}

__PACKAGE__->meta->make_immutable;
1;
